### DISCUSSION TITLE
Seeking guidance on the migration path for changing the csi.kubeletRootDir configuration value on an active cluster. Without downtime.

### DISCUSSION THREAD

**Discussion by larssb** [longhorn/longhorn#5593] [Q&A]
Using the [Longhorn Helm Chart](https://github.com/longhorn/longhorn/tree/master/chart) one have the option of configuring the [`csi.kubeletRootDir`](https://github.com/longhorn/longhorn/blob/master/chart/values.yaml#L101) ... I want to do this on several already live K3s v1.24+ clusters, as we're changing to dedicating an LVM2 for container related data ( following this [approach/guide](https://github.com/k3s-io/k3s/issues/2068#issuecomment-1374672584) ) and therefore need to point `Longhorn` to where the `kubelet` now "lives".

Via `helm` we can update the value to `csi.kubeletRootDir`. That action updates the `longhorn-driver-deployer`. However, the `longhorn-csi-plugin`, `csi-attacher`, `csi-provisioner`, `csi-resizer` and the `csi-snapshotter` also needs to "know" of the change to `csi.kubeletRootDir`. The best way I've found to communicate the change to these workloads, have been to edit them directly. This then gives them the updated value of `csi.KubeletRootDir` as they're re-created/re-deployed.

However, we then need to remove "old" `worker` `nodes` from the cluster by draining, then `kubectl delete ...` them ( as they where **NOT** bootstrapped with the LVM2 for container data ).  This will cause workloads consuming `Longhorn Volumes` to of course try to `umount` and `detach` as they're drained from the "old" nodes to go to the new ones.
However, they can't successfully complete the `umount` cmd's as the `kubelet dir` changed and the `csi.sock` is therefore no longer where the `Longhorn instance-i-* and instance-r-*` `workloads` expects them to be.

Looking into `Longhorn` logs I found:

```txt
2023-03-16 20:22:19 Mar 16 20:22:19 cluster-name-worker-14 k3s[1163]: E0316 20:22:19.4789211163
nestedpendingoperations.go:348] Operation for"{  volumeName:kubernetes.io/csi/driver.longhorn.io^pvc-9dc67125-7cd4-4c2c-
ac3d-1446936e4305  podName:23e33d25-01da-4faa-975a-71f1fe03a8b9 nodeName:}"failed. No retries permitted until 2023-03-16 
20:24:21.478872883 +0100 CET m=+641131.112565477(durationBeforeRetry 2m2s).Error:
UnmountVolume.TearDown failed for volume "alertmanager-cluster-name-kube-prom-alertmanager-db"
(UniqueName: "kubernetes.io/csi/driver.longhorn.io^pvc-9dc67125-7cd4-4c2c-ac3d-1446936e4305")pod "23e33d25-01da-4faa-975a-71f1fe03a8b9"
(UID: "23e33d25-01da-4faa-975a-71f1fe03a8b9") :kubernetes.io/csi:
Unmounter.TearDownAt failed to get CSI client:driver name driver.longhorn.io not found in the list 
of registered CSI drivers
```

This can cause `Attach` errors when the workloads have migrated to newly introduced `worker nodes`. E.g. the `pendingNodeID` on `Longhorn Volumes` have such `Attach` issues, is an ID of a now removed "old" `worker node`.

> However, the `alertmanager-cluster-name-kube-prom-alertmanager-db` `workload` from the err above actually never got any `Attach issues`. Others did however - in their case I'm unsure if they "saw" the same `Unmounter.TearDownAt failed to get CSI client:driver name driver.longhorn.io not found in the list of registered CSI drivers` error - but I would believe so. They ran on the same `worker` - it's likely just that I didn't spot the same error for those specific workloads in the `Longhorn` log.

In all fairness it should of course be mentioned that I did not follow best practices on [`Removing a Node`, in the `Longhorn` _docs_](https://longhorn.io/docs/1.4.1/volumes-and-nodes/maintenance/#removing-a-node). As I simply wasn't aware of these before a couple of days ago. The "old" `worker nodes` where simply `kubectl drained ...` >> then `kubectl deleted ...`.
So this might have been partly or fully the reason for the attach errors - and not the theory I have on the change to the `csi.kubeletRootDir` and the re-creation/re-deployment of the above `csi*` and `longhorn-driver-deployer` `Lonnghorn workloads` - making the `csi.sock` not being in the expected location for the `Longhorn instance-i-* and instance-r-*` `workloads` ... I'm not sure :-(

---

Because of these `Attach` errors and because I would really really prefer to not `scale` down `StatefulSet` to avoid downtime, I'm seeking guidance on the migration path to take on live clusters to achieve dedicating container related data on an LVM2 disk and the consequence of the change to the `--kubelet-arg=root-dir=....` necessary for that to happen.

Thank you very much.

**Comment by larssb** [comment]
I've been reading the Cilium v1.13 [release blog](https://isovalent.com/blog/post/cilium-release-113/) and in there they mention a custom resource named `CiliumNodeConfig`. If `Longhorn` had the same thing I could gradually introduce nodes with the `csi.kubeletRootDir` change.

Is this an idea that's viable for the `Longhorn` project?

Thank you.

**Comment by weizhe0422** [comment]
Longhorn relies on a socket file which is under kubelet dir to register the Longhorn CSI plugin, then kubelet can interact with Longhorn.

IIUC, you had modified the `kubeletRootDir`, then you want to migrate the volume from the old node to the new node with perform reattach. The pods on the old node were already using the new kubeletRootDir path but CSI plugin was still using the old path. And it makes the detach operation not succeeded. 

Since Longhorn cannot apply different kubelet root dirs to the pods of one deployment on different nodes. Therefore, it seems like need to find a workaround way:
1. detach the volume manually and let the kubelet knows the volume was detached, e.g., deleting the corresponding VolumeAttachement object
2. Or help launch correct CSI plugin pods on old nodes.

**Comment by larssb** [comment]
Okay thanks Ray. How do I do number two option? One will of course bring some downtime but not much …. Might be the best option for us.What do you think? How do we approach this from here?Thank you very much.--Mvh. / Kind regardsLars Bengtsson- Health is wealth, movement is medicine -On 2 May 2023, at 18.56, Ray Chang ***@***.***> wrote:﻿
Longhorn relies on a socket file which is under kubelet dir to register the Longhorn CSI plugin, then kubelet can interact with Longhorn.
IIUC, you had modified the kubeletRootDir, then you want to migrate the volume from the old node to the new node with perform reattach. The pods on the old node were already using the new kubeletRootDir path but CSI plugin was still using the old path. And it makes the detach operation not succeeded.
Since Longhorn cannot apply different kubelet root dirs to the pods of one deployment on different nodes. Therefore, it seems like need to find a workaround way:

detach the volume manually and let the kubelet knows the volume was detached, e.g., deleting the corresponding VolumeAttachement object
Or help launch correct CSI plugin pods on old nodes.

**Comment by LarsBingBong** [comment]
@weizhe0422 any feedback on my question?
